FWCore/Framework/src/EventProcessor.cc

```diff
@@ -2203,64 +2205,87 @@ namespace edm {
   }
 
   void EventProcessor::handleNextEventForStreamAsync(WaitingTaskHolder iTask, unsigned int iStreamIndex) {
+    auto handleNextEventStatus =
+        std::make_shared<std::atomic<unsigned int>>(static_cast<unsigned int>(HandleNextEventStatus::kRunning));
+
     auto group = iTask.group();
-    sourceResourcesAcquirer_.serialQueueChain().push(*group, [this, iTask = std::move(iTask), iStreamIndex]() mutable {
-      CMS_SA_ALLOW try {
-        auto status = streamLumiStatus_[iStreamIndex].get();
-        ServiceRegistry::Operate operate(serviceToken_);
+    sourceResourcesAcquirer_.serialQueueChain().push(
+        *group, [this, iTask, iStreamIndex, handleNextEventStatus]() mutable {
+          // There is a bit of code after the push call that might run
+          // streamEndLumiAsync before this functor runs. If that happened
+          // just return and do nothing. Before we check to see if it happened
+          // we wait and give that code time to finish. I expect it to be
+          // infrequent that we actually have to wait at this point.
+          while (handleNextEventStatus->load() == static_cast<unsigned int>(HandleNextEventStatus::kRunning)) {
```

Comment:
I'm concerned about holding onto a shared resource with such a wait loop.

Reply:
If you need to sequence this, then why not move 
```
if (streamLumiStatus_[iStreamIndex]->haveStartedNextLumiOrEndedRun()) {
```
block to before the pushing into the ` sourceResourcesAcquirer_.serialQueueChain()`? 

This will give almost identical results to what you have but does not require the `handleNextEventStatus`. The only advantage to your change is if during the very tiny amount of time it takes to push the task into the serial queue another stream happens to finish that Lumi. Given the push is probably taking less than 1us, I don't think the added complication of `handleNextEventStatus` makes sense.

Reply:
The problem with what you are suggesting is exactly as you described. If the conditional returns false, then it will put the task in the serial source queue. I agree that is very fast. I agree for it to be a problem another thread would need to execute global begin lumi and stream begin lumi and ask the source what is next before the task is put in the serial source queue. The probability of that happening is low. Probably it would require the operating system pausing the thread in that tiny time window. On the other hand, there is not a hard requirement to prevent that from occurring. The consequence would be exactly what this PR is supposed to avoid, a delay in closing lumi N until we hit lumi N+2. It's hard to evaluate the probability. Is it once in a million years so that it doesn't happen in the lifetime of CMS or once a year. I don't how to evaluate that. What happens if every once in a long while the lumi closing is delayed? Does that cause the process to run out of memory and crash? I don't know the consequence of that either. If it's not a big deal if it occurs rarely then you are definitely right.

The advantage of the approach in the current version of the PR is that things will be sequenced to avoid the late lumi closings 100% of the time. Reversing the order and checking the condition after inserting into the queue causes the task in the serial queue to always be ahead of any potentially blocking task. The conditional in the spin wait will also rarely fail. The competition there starts from the moment the task is placed in the serial task queue. On that thread, it reads the value of one atomic variable and compares it to a constant, then an if branch, and then it sets the value of another atomic. If it gets that far, the wait condition always fails and there is no wait. Also that will end the spin wait in the unlucky event that we're spinning. That is competing against another thread being available and starting to execute the task that was in the serial queue and getting to the first line of that functor. The worst consequence is that there is a spin wait for a short time. The lumi closing delay never occurs.

It is a little more complicated. 12 lines versus 3 lines.

There is an extra atomic value needing to be created, set, and checked. It doesn't live very long. One shared_ptr that goes out of scope when handleNextEventForStreamAsync returns. Another one that goes out of scope when the functor in the task from the serial queue returns.

I had it coded the way you suggested first. It bothered me that it could have the wrong behavior sometimes.

Reply:
For the following question I shorten the code in question as
```cpp
sourceResourcesAcquirer_.serialQueueChain().push([](){
  if (...) { ... 
  } else {
    ...
   streamEndLumiAsync(iTask, iStreamIndex); // A
  } 
});
if (...) {
  ...
  streamEndLumiAsync(std::move(iTask), iStreamIndex); // B
};
```
Do the parts in A and B need to be sequenced, i.e. the code in the lambda pushed to the source's serial task queue must happen before the code in the if-block surrounding B, or vice versa?

Or, would be sufficient that either A or B call the `streamEndLumiAsync()`?

I'm wondering if a simple compare-and-swap to decide whether A or B calls `streamEndLumiAsync()` would be sufficient?

Reply:
I actually tried that version also. That was my second version of this. When I ran the tests I realized that if it calls streamEndLumiAsync at B, then when it hits the functor from the stream queue later the lumi might have already ended and the lumi status object (which the functor uses) might no longer exist and then a seg fault occurs. At the very beginning of the functor it needs to know that streamEndLumiAsync has already been called and it must immediately return and do nothing. There might be some way around that but I am not sure it would be any simpler than what is there now.

Reply:
> At the very beginning of the functor it needs to know that streamEndLumiAsync has already been called and it must immediately return and do nothing.

Does it need to know exactly that `streamEndLumiAsync()` has been already called, or would it be sufficient to know if `streamEndLumiAsync()` will be called "soon" by another thread (and therefore skip the rest of the functor)?

Reply:
My thought at this point is that we should discuss this in person on Monday. Hopefully all three of us are on site.

Chris's suggestion is the simplest. It is quite possible the behavior I am worried about is so improbable that it does not occur in the lifetime of CMS. The simple approach would certainly make the problem reported by the online occur much less often.

On the other hand, there is a way to completely avoid it and it seems to me the complexity and overhead aren't very significant. I hate trying to debug failures that are hard to reproduce and rare. If this happened we'd probably never reproduce it and just give up trying to debug it after a while. And we are trying to address a problem that was noticed and reported by online folks. This isn't a theoretical issue that might arise.

Reply:
> My thought at this point is that we should discuss this in person on Monday. Hopefully all three of us are on site.

Ok, let's do that (I agree, likely easier). I'm planning to come on site.

Reply:
(Crossing emails. I didn't see you last one before I sent my last reply)

Knowing that it will be called soon is good enough.

I am not quite sure exactly what you have in mind and maybe the following are not relevant, but there are the slightly subtle points that for the first stream it needs to always go through the functor and it must know that for the same stream it is going to be called soon.